docs/proposals/ktl.md

```diff
@@ -0,0 +1,202 @@
+
+#### Configuration
+
+Use [viper](https://github.com/spf13/viper) to configure dispatch precedence
+
+### Cli commands
+
+All new cli commands should be built outside of kubernetes/kubernetes and vendored into the kubernetes/ktl.  Command repos
+implement an interface that returns a list of cobra (sp13) commands, which ktl can register under the root.
+
+Initially new cli commands can be built in the kubernetes/kubectl, but if needed, development maybe moved to other repos.
+
+#### Conventions
+
+- ktl top level commands are one of
+  - command groups that work against a specific API group/versions (e.g. isto, svcat, kube)  e.g. `create`
+    for a given resource.
+  - generic commands that are agnostic to specific APIs, but instead discover API metadata and work against
+    all APIs following published conventions.  e.g. a `scale` that works against anything with a scale sub resource
+
+
+### Library infrastructure
+
+Develop shared set of client / cli libraries that handle registering flags / environment variables / config files
+and provide common functionality to cli clients of Kubernetes.
+
+- reading and parsing kubeconfig
+- printing libraries
+- reading and parsing config into objects from files, directories, stdin
+- indexing + querying discovery API and openapi data
+- manipulating unstructured objects using openapi schema
+- merging objects
+- generating patches from objects
+- writing objects to the server - optimistic locking failure retries with exp backoff
+- providing tty for subresources that accept - e.g. exec, attach
+- defining shared exit codes (e.g. exit code when not doing a conditional write)
+- test infrastructure for unit and integration tests
+  - support version skew testing
+
+### Build and release infrastructure
+
+Develop build triggers to automatically cut and publish builds based on the presence of GitHub tags.  Aggregate
+release notes from vendored commands.
+
+- use GCP container builder + mirrored GCP source repo
+- publish binary to gs:// bucket
+- publish binary + release notes to GitHub releases page
+
+## Alternatives considers
+
+*Don't vendor in commands, make them plugins instead*
+
+- This is a more complicated approach that can be considered in a later iteration.
+
+*Name `ktl` -> `kubectl` instead and attach `kubectl` subcommands directly to the root so
+that users don't see a different interface*
+
+- Many of the commands should be rewritten and have the behavior slightly modified to be more correct.
+  Renaming the binary allows us to expose the new command under a different name - avoiding breaking compatibility
+- This also allows us to more easily refine the design and architecture by
+  - restructuring the command groupings
+  - dropping commands we want to phase out
+
+But what about all the old blog posts, docs, books?
+
+- These become out of date anyway as new and improved solutions are built (e.g. new APIs are introduced)
+- Docs can be updated
```

Comment:
Does it make sense to add *Migration* section which talks about how people can start using new CLI without disruption their existing tools. For ex.
Aliasing "kubectl" --> "ktl kube"